lite/examples/sound_classification/ios/SoundClassification/TFLAudioRecord/TFLAudioRecord.m

```diff
@@ -108,78 +115,105 @@ - (void)startTappingMicrophoneWithCompletionHandler:
                                     description:
                                         @"You have passed an unsupported number of channels."];
-                          completionHandler(nil, frameBufferFormatError);
-                        } else if (![self->_ringBuffer loadWithBuffer:floatBuffer
-                                                               offset:0
-                                                                 size:floatBuffer.size
-                                                                error:&loadError]) {
-                          completionHandler(nil, loadError);
-                        } else {
-                          GMLFloatBuffer *outFloatBuffer = [self->_ringBuffer.buffer copy];
-                          completionHandler(outFloatBuffer, nil);
-                        }
+                          self->globalError = frameBufferProcessingError;
+                        } else if (![self->_ringBuffer loadBuffer:floatBuffer
+                                                          offset:0
+                                                            size:floatBuffer.size
+                                                           error:&frameBufferProcessingError]){
+                            self->globalError = frameBufferProcessingError;
+                          }
+                          else {
+                            self->globalError = nil;
+                          }
                         break;
                       }
-                      case AVAudioConverterOutputStatus_Error: {
-                        completionHandler(nil, conversionError);
+                      case AVAudioConverterOutputStatus_Error: // fall through
+                      default: {
+                        if (!conversionError) {
+                          [TFLUtils createCustomError:&conversionError withCode:TFLAudioErrorCodeAudioProcessingError description:@"Some error occurred during audio processing"];
+                        }
+                        self->globalError = conversionError;
                         break;
                       }
-                      default:
-                        completionHandler(nil, nil);
-                        break;
                     }
                   });
                 }];
 
-  NSError *engineStartError = nil;
 
   [_audioEngine prepare];
-  [_audioEngine startAndReturnError:&engineStartError];
+  [_audioEngine startAndReturnError:error];
 
-  if (engineStartError) {
-    completionHandler(nil, engineStartError);
-  }
 }
 
-- (void)checkPermissionsAndStartTappingMicrophoneWithCompletionHandler:
-    (void (^)(GMLFloatBuffer *_Nullable buffer, NSError *_Nullable error))completionHandler {
+- (void)startRecordingWithCompletionHandler:
+    (void (^)(NSError *_Nullable error))completionHandler {
   [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```

Comment:
As this call requires user to add a permission description string to the Info.plist, I think we should let the app developer get the permission in their app, outside of the scope of the Task Library. This will be more aligned to the Android API. (Python doesn't have the concept of permission request).

Could you change the API as follow?
- [Check if the app has recording permission](https://developer.apple.com/documentation/avfaudio/avaudiosession/1616463-recordpermission).
   - If it's denied, return a "permission denied" NSError.
   - If it's undetermined, return an NSError with the error message suggesting the user to read the Apple's requestRecordPermission documentation and request permission.
- Change the startRecordingWithCompletionHandler method to startRecordingWithError as we don't async interface anymore.

https://developer.apple.com/documentation/avfaudio/avaudiosession/1616463-recordpermission

Reply:
This is fixed now.